homeassistant/components/whirlpool/sensor.py

```diff
@@ -0,0 +1,180 @@
+"""The Washer/Dryer Sensor for Whirlpool Appliances."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+
+from whirlpool.washerdryer import MachineState, WasherDryer
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import STATE_UNKNOWN
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.typing import StateType
+
+from . import WhirlpoolData
+from .const import (
+    CYCLE_FUNC,
+    DOMAIN,
+    ICON_D,
+    ICON_W,
+    MACHINE_STATE,
+    TANK_FILL,
+    WASHER_STATE,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def washer_state(washer: WasherDryer) -> str | None:
+    """Determine correct states for a washer."""
+
+    machine_state = washer.get_machine_state()
+    machine_cycle = None
+    for count, func in zip(range(len(CYCLE_FUNC)), CYCLE_FUNC):
+        if func(washer):
+            machine_cycle = WASHER_STATE.get(count)
+            break
+
+    if washer.get_attribute("Cavity_OpStatusDoorOpen") == "1":
+        machine_cycle = "Door open"
+        return machine_cycle
+
+    if machine_state == MachineState.RunningMainCycle and machine_cycle:
+        return machine_cycle
+
+    return MACHINE_STATE.get(machine_state, STATE_UNKNOWN)
+
+
+@dataclass
+class WhirlpoolSensorEntityDescriptionMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable
+
+
+@dataclass
+class WhirlpoolSensorEntityDescription(
+    SensorEntityDescription, WhirlpoolSensorEntityDescriptionMixin
+):
+    """Describes Whirlpool Washer sensor entity."""
+
+
+SENSORS: tuple[WhirlpoolSensorEntityDescription, ...] = (
+    WhirlpoolSensorEntityDescription(
+        key="state",
+        name="state",
+        entity_registry_enabled_default=True,
+        icon=ICON_W,
+        has_entity_name=True,
+        value_fn=washer_state,
+    ),
+    WhirlpoolSensorEntityDescription(
+        key="timeremaining",
+        name="time remaining",
+        device_class=SensorDeviceClass.DURATION,
+        state_class=SensorStateClass.MEASUREMENT,
+        native_unit_of_measurement="s",
+        entity_registry_enabled_default=True,
+        icon=ICON_W,
+        has_entity_name=True,
+        value_fn=lambda WasherDryer: WasherDryer.get_attribute(
+            "Cavity_TimeStatusEstTimeRemaining"
+        ),
+    ),
+    WhirlpoolSensorEntityDescription(
+        key="DispenseLevel",
+        name="Dispense Level",
+        device_class=SensorDeviceClass.VOLUME,
```

Comment:
Reading the test, this sensor returns a value like "50%" however looking at https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes this is not one of the supported units. Am I understanding that right?

Reply:
I was looking at other sensors for % and they seem to set `native_unit_of_measurement=PERCENTAGE`

Reply:
Well... it is volume and it is % full.... More like a battery but not really.
It actually returns % except sometimes it reports unknown, empty, or "active" - these were pulled from the whirlpool UI.  Maybe it should just be no device class? Looks like device_class is just to workout conversions between unit types anyway.

Reply:
I think of it this way: What is the standard way that sensors should report a percentage to home assistant, such that the system knows what the value is. This can help with other things such as representing in the UI properly, etc.  How the data is returned by the APIs, libraries, etc is an implementation details handled by the integration.

Hopefully the motivations makes sense as its not arbitrary busy work :) but meant to make it easier to build new features on top of or integrate with other parts of the system. (e.g. imagine in the future something that makes it easier to for users to configure an alert when the detergent is low, similar to how we might alert when battery is low, etc)

Reply:
Totally makes sense.  I'm not sure what to do with the other states if I moved to a pure %. 

https://github.com/mkmer/HomeAssistant_Core/blob/Whirlpool-laundry/homeassistant/components/whirlpool/sensor.py#L28-L35

At times the appliance just reports "active" when it for some reason doesn't know the level.  Other times it can report "unknown".  Empty could just be 0%, but I'm not sure what I would use in the world of % for "Unknown" and "Active"
It's probably best to have no device_class for this one....

Reply:
Yeah, i agree with you that it makes sense to have no device class. it seems like its super specific to the start of the wash cycle and not something general, and custom to this device.